File: Commands/Slash/Ban.cs
using Adramelech.Configuration;
using Adramelech.Extensions;
using Adramelech.Utilities;
using Discord;
using Discord.Interactions;
using Discord.WebSocket;

namespace Adramelech.Commands.Slash;

public class Ban : InteractionModuleBase<SocketInteractionContext<SocketSlashCommand>>
{
    [SlashCommand("ban", "Ban a member")]
    [RequireUserPermission(GuildPermission.BanMembers)]
    [RequireBotPermission(GuildPermission.BanMembers)]
    [RequireContext(ContextType.Guild)]
    public async Task BanAsync([Summary("user", "The user to ban")] SocketGuildUser user,
        [Summary("reason", "The reason for the ban")]
        string reason = "No reason provided",
        [Summary("prune_days", "The number of days to prune messages")]
        int pruneDays = 0,
        [Summary("ephemeral", "Whether the response should be ephemeral")]
        bool ephemeral = false)
    {
        if (user.Id == Context.User.Id)
        {
            await Context.SendError("You can't ban yourself!");
            return;
        }

        if (user.Hierarchy >= Context.Guild.CurrentUser.Hierarchy)
        {
            await Context.SendError("You cannot ban a member with a higher or equal role than you!");
            return;
        }

        await user.BanAsync(pruneDays, reason);

        await RespondAsync(
            embed: new EmbedBuilder()
                .WithColor(Config.EmbedColor)
                .WithTitle("Member Banned")
                .WithDescription($"User {user.Username} has been banned")
                .AddField("Reason", $"`{reason}`")
                .AddField("Author", Context.User.Mention)
                .Build(),
            ephemeral: ephemeral);

        if ((await ErrorUtils.TryAsync(() =>
                user.SendMessageAsync($"You have benn banned from {Context.Guild.Name}. Reason: {reason}"))).IsFailure)
            await Context.SendError("Failed to notify the user about the ban");
    }
}

File: bot/src/commands/slash/ban.ts
import type { Command } from '~/types/command.ts';
import {
  type ChatInputCommandInteraction,
  Colors,
  ComponentType,
  InteractionContextType,
  MessageFlags,
  PermissionFlagsBits,
  SlashCommandBuilder,
  userMention,
} from 'discord.js';
import { sendError } from '~/utils/sendError.ts';
import config from '~/config.ts';
import { stripIndents } from 'common-tags';
import { fromAsyncThrowable } from 'neverthrow';

export const command = <Command>{
  data: new SlashCommandBuilder()
    .setName('ban')
    .setDescription('Ban a member')
    .addUserOption((option) =>
      option
        .setName('user')
        .setDescription('The user to ban')
        .setRequired(true),
    )
    .addStringOption((option) =>
      option.setName('reason').setDescription('The reason for the ban'),
    )
    .addIntegerOption((option) =>
      option
        .setName('prune-days')
        .setDescription('The number of days to prune messages'),
    )
    .addBooleanOption((option) =>
      option
        .setName('ephemeral')
        .setDescription('Whether to show the response only to you'),
    )
    .setContexts(InteractionContextType.Guild)
    .setDefaultMemberPermissions(PermissionFlagsBits.BanMembers),
  async execute(intr: ChatInputCommandInteraction) {
    const user = intr.options.getUser('user', true);
    const reason = intr.options.getString('reason') ?? 'No reason provided';
    const pruneDays = intr.options.getInteger('prune-days') ?? 0;
    const ephemeral = intr.options.getBoolean('ephemeral') ?? false;

    if (user.id === intr.user.id)
      return await sendError(intr, "You can't ban yourself");
    if (user.id === intr.client.user.id)
      return await sendError(intr, "You can't ban me OwO");
    if (user.id === intr.guild?.ownerId)
      return await sendError(intr, "You can't ban the server owner");

    const member = intr.guild!.members.cache.get(user.id)!;
    const botMember = intr.guild!.members.cache.get(intr.client.user.id)!;
    if (
      member.roles.highest.comparePositionTo(botMember.roles.highest) >= 0 ||
      !member.bannable
    )
      return await sendError(intr, "I can't ban this user");

    const banResult = await fromAsyncThrowable(
      () =>
        member.ban({
          reason,
          deleteMessageSeconds: pruneDays * 86400,
        }),
      (e) => `Failed to ban user:\n${String(e)}`,
    )();
    if (banResult.isErr()) return await sendError(intr, banResult.error);

    await intr.reply({
      flags: ephemeral
        ? MessageFlags.IsComponentsV2 | MessageFlags.Ephemeral
        : MessageFlags.IsComponentsV2,
      components: [
        {
          type: ComponentType.Container,
          accent_color: config.EMBED_COLOR,
          components: [
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              # Member banned
              User \`${user.username}\` has been banned
              ### :warning: Reason
              \`\`\`${reason}\`\`\`
              `,
            },
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              ### :shield: Author
              ${userMention(intr.user.id)}
              `,
            },
          ],
        },
      ],
    });

    const notifyResult = await fromAsyncThrowable(
      () =>
        user.send({
          flags: MessageFlags.IsComponentsV2,
          components: [
            {
              type: ComponentType.Container,
              accent_color: Colors.Red,
              components: [
                {
                  type: ComponentType.TextDisplay,
                  content: stripIndents`
                  # You have been banned
                  ### Guild
                  \`\`\`${intr.guild?.name}\`\`\`
                  `,
                },
                {
                  type: ComponentType.TextDisplay,
                  content: stripIndents`
                  ### Reason
                  \`\`\`${reason}\`\`\`
                  `,
                },
              ],
            },
          ],
        }),
      (e) => `Failed to notify the user about the ban:\n${String(e)}`,
    )();
    if (notifyResult.isErr()) return await sendError(intr, notifyResult.error);
  },
};
